usb: netusb add CDC EEM usbnet function

loicpoulain opened the pull request:
This PR adds CDC Ethernet Emulation Model function driver (CDC EEM). This usb network
function can be used for ethernet over USB. Ethernet packet are encapsulated within EEM packets. An EEM pkt contains 2-byte header and 4-byte sentinel (or crc).

Contrary to CDC RNDIS and CDC ECM, this class requires only two endpoints (bulk in/out).

pfalcon commented:
@loicpoulain : What's the status of this? Can you please add description of this PR?

pfalcon commented:
This is all of: marked "in progress", "1.11", adds a new feature, and in conflict state. 1.11 is in pre-release freeze now, so new features aren't accepted. Please let me change milestone to 1.12 then.

loicpoulain commented:
This PR needs to be reworked and mainly depends on PR #5983...

Commit:
usb: netusb: Rework netusb media connect/disconnect

Consider media connected when interface is selected by the host and
disconnected on device diconnection.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
usb: netusb: Add CDC EEM network usb function

Add CDC Ethernet Emulation Model function driver. This usb network
function can be used for ethernet over USB. Ethernet packet are
encapsulated within EEM packets. An EEM pkt contains 2-byte header
and 4-byte sentinel (or crc).

Note that EEM packets can be split across USB packets but shall not
be split across USB transfers.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

jfischer-phytec-iot commented on include/usb/usb_common.h:
```diff
@@ -105,6 +105,7 @@
 /* Sub-classes */
 #define ACM_SUBCLASS			0x02
 #define CDC_ECM_SUBCLASS		0x06
+#define CDC_EEM_SUBCLASS		0x0c
```
remove this line, it is already defined in include/usb/class/usb_cdc.h

jfischer-phytec-iot commented on include/usb/usb_common.h:
```diff
@@ -125,6 +126,7 @@
 #define ACM_VENDOR_PROTOCOL		0xFF
 /* Misc protocols */
 #define MISC_ETHERNET_PROTOCOL		0x01
+#define EEM_PROTOCOL			0x07
```
also here, it is already defined in include/usb/class/usb_cdc.h

jfischer-phytec-iot commented on subsys/usb/class/netusb/function_eem.c:
```diff
@@ -0,0 +1,167 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define NET_LOG_ENABLED 1
```
It does not have to be enabled

jfischer-phytec-iot commented on subsys/usb/class/netusb/function_eem.c:
```diff
@@ -0,0 +1,167 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define NET_LOG_ENABLED 1
+#define SYS_LOG_LEVEL 3
```
s/3/CONFIG_SYS_LOG_USB_DEVICE_NETWORK_DEBUG_LEVEL

finikorg reviewed (changes requested):
Please also add sanity_check build configuration so we could catch build errors

loicpoulain commented on subsys/usb/class/netusb/Kconfig:
```diff
@@ -23,6 +23,12 @@ config USB_DEVICE_NETWORK_ECM
+	prompt "Ethernet Emulation Model (EEM) Networking device"
+	select USB_DEVICE_NETWORK
+	default n
+
```
Sure.

loicpoulain commented on subsys/usb/usb_descriptor.c:
```diff
@@ -598,6 +608,61 @@ static struct dev_common_descriptor common_desc = {
+			.bInterfaceCount = 0x01,
+			.bFunctionClass = COMMUNICATION_DEVICE_CLASS,
+			.bFunctionSubClass = CDC_EEM_SUBCLASS,
+			.bFunctionProtocol = 0,
```
thanks.

Commit:
samples: net: Add test cases for USB EEM

Add test for zperf and http_server apps.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

jukkar reviewed (approved):
LGTM

jli157 commented:
@loicpoulain I know this function was merged two years ago. Do we have an introduction how to use EEM stack from both the host and device sides? It seems EEM is quite different from other two, and raw socket in zephyr doesn't work with it. It is appreciated if an introduction or detailed example is provided. Thank you!

loicpoulain commented:
> @loicpoulain I know this function was merged two years ago. Do we have an introduction to how to use EEM stack from both the host and device sides?

There a lot of network samples that can be used with CONFIG_USB_DEVICE_NETWORK_EEM, however, I agree it would be nice to have a sample specifically configured with this function.

>  It seems EEM is quite different from other two,

No, it's quite similar to RNDIS or ECM.

> and raw socket in zephyr doesn't work with it.

How did you test that? I've just tested echo-server sample with EEM on disco_l475_iot1 without any problem.

jli157 commented:
> > @loicpoulain I know this function was merged two years ago. Do we have an introduction to how to use EEM stack from both the host and device sides?
> 
> There a lot of network samples that can be used with CONFIG_USB_DEVICE_NETWORK_EEM, however, I agree it would be nice to have a sample specifically configured with this function.
> 
> > It seems EEM is quite different from other two,
> 
> No, it's quite similar to RNDIS or ECM.
> 
> > and raw socket in zephyr doesn't work with it.
> 
> How did you test that? I've just tested echo-server sample with EEM on disco_l475_iot1 without any problem.

I'm a bit confused by the design purpose of EEM. From the USB EEM protocol spec, it says "Unlike CDC ECM, EEM does not extend an interface across the USB bus but instead considers the USB bus to be a vehicle for moving Ethernet packets". So, what makes EEM different from other two? How can we use EEM to get that benefit?

loicpoulain commented:
> I'm a bit confused by the design purpose of EEM. From the USB EEM protocol spec, it says "Unlike CDC ECM, EEM does not extend an interface across the USB bus but instead considers the USB bus to be a vehicle for moving Ethernet packets". So, what makes EEM different from other two? Compared with other usb network protocols, what benefit can we get by using this one?

Well ECM/RNDIS are normally used to drive a real network controller over USB (e.g. USB to ethernet adapter):
 HOST(iface)===USB===>Ethernet-Controller===ETHERNET====>
EEM is just a simple protocol to do ethernet over USB:
 HOST(iface)===USB===>(iface)DEVICE

Though in Zephyr ECM and RNDIS have been implemented to act as EEM and just offer a connection to Zephyr internal network stack (a virtual interface is created with mac address...).

EEM is a simpler protocol and only use two endpoints (instead of 3 for ECM and RNDIS), which can be useful on device with limited number of endpoints. It is also possible to send multiple ethernet packets in one USB transfer (instead of one ethernet packet per USB transfer) which in theory improves the maximum throughput.

jli157 commented:
Got it, thanks for the explanation!
